### ISSUE TITLE
Run build script that has devDependencies after push to remote repo

### ISSUE THREAD

**Issue by tstringer** [projectkudu/kudu#2239]
## Environment
- Node.js web app
- Azure App Service target
- Git repo (App Service CD)

## Background Story
I have a build script that generates the build/dist files. Ultimately these dist files should *not* live in the repo, but they should be generated after pushing to the git repo in Azure App Service. The catch here is that the build script has dependencies on `devDependencies` and it seems like the default kudu behavior is to run `npm install --production` therefore not installing the `devDependencies`. I can verify this in the kudu management web portal for the site (SCM).

## Desire
I'd like to have kudu run a normal `npm install` and then subsequently run this build script (i.e. `npm run build`) after I push to the kudu repo in Azure App Service.

## What I've Tried
I created a `.deployment` file in the repo root and put these contents in this file...

```
[config]
command = npm install
```

After I run `git push azure master` I can see that the remote says `Running custom deployment command` and a bunch of npm warnings that I would normally see when running `npm install` locally. Unfortunately this seems to indefinitely hang. I've waiting quite a bit of time and the remote never tells me it is done. Doing `npm install` locally definitely doesn't take nearly this long.

What is the best way to accomplish what I'm attempting here?

Thank you in advance!

**Comment by tstringer** [comment]
**Update**: I was able to generated a deployment script by running `azure site deploymentscript --node`. I changed the npm install command and removed `--production`. This seemed like it was working, but then I ran into the max path issue on Windows.

I tried to do the same with App Service on Linux, but it seems like that doesn't like my shell script:

> remote: /bin/bash: /bin/bash: cannot execute binary file

**Comment by davidebbo** [comment]
What npm version are you using? 3.x is smarter about path length and can avoid some of the issues.

Let's discuss Linux separately so we don't turn this issue into a multi-headed monster :)

**Comment by tstringer** [comment]
@davidebbo 😄  totally agree and thanks for the quick response! In SCM console `npm -v` shows me I'm running `2.15.8`. What's the best way to force npm v3?

**Comment by tstringer** [comment]
@davidebbo perfect! So that was the first step in a few I ended up doing. `WEBSITE_NODE_DEFAULT_VERSION` setting to 6.9.1 forced npm v3, which got me around the max path issue. For some reason, in my `deploy.sh` it didn't like me trying to run `npm run build` to run my npm build script. It reverted to a *very* old version of node/npm (seemed to resolve to `d:\program files (x86)\nodejs\node.exe` which is verified as being an extremely old build).

The workaround here was to remove the build script execution from `deploy.sh` and use an npm script `postinstall` to then run my build script. Voila! Worked like a charm.

Thanks so much for the help, David! 🚀

**Comment by premchandrasingh** [comment]
I know this issue is closed but I still having the issue even after removing `--production` flag. The npm version is 3.10.8 and node version is 6.9.1. 

What I am trying to do is to run bower install and run gulp task but installing bower and gulp is failing which I defined in devDependancies. What I might be missing?

Thanks in advance.

**Comment by tstringer** [comment]
@premchandrasingh is your code open source? If so, I'd be happy to take a look. I also recommend you walk through [the blog post I wrote on this very topic](https://medium.com/@trstringer/custom-build-logic-post-git-push-with-azure-app-service-and-kudu-for-a-node-js-web-app-1b2719598916#.2884kfn2o) which steps through all requirements to get this working.

And if your project is not open source, would you be able to share your effect `deploy.sh` script?

**Comment by premchandrasingh** [comment]
@tstringer Thanks for your quick response. My code is not open source but I separated `deploy.cmd` only into a public repository. Here is the code which has issue https://github.com/premchandrasingh/kuduscript/commit/328bed6ef2983596dac522562c63f8033090f741

By the way, I have figured out some hack around using `--only=dev` flag and this is kinds of what I like. Here is my working script https://github.com/premchandrasingh/kuduscript/commit/9b94c0c9aa763f7bfc292665a213eddd29ec1111

Any suggestions are welcome
